### PR TITLE
Added initial network profile powershell commands and ServiecAssociationLink to Subnet Model

### PR DESCRIPTION
**Pull request by Flip-Liquid**
## Description

Exposing new commands for manipulating network profile:

- 'Get-AzureRMNetworkProfile',
- 'New-AzureRMNetworkProfile',
- 'Remove-AzureRMNetworkProfile',
- 'Set-AzureRMNetworkProfile',
- 'New-AzureRMNetworkProfileContainerNetworkInterfaceConfig',
- 'New-AzureRmNetworkProfileContainerNetworkInterfaceConfigIpConfig',

Also adding service association link property on subnet

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr) [[Relevant review here]](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/159)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

## Design review
[Cmdlet design review #162](https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/162)

### CODE REVIEW

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.Netcore.csproj [hunk 1]
```diff
@@ -25,9 +25,10 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.Azure.Graph.RBAC" Version="3.4.0-preview" />
-    <PackageReference Include="Microsoft.Azure.Management.Network" Version="19.4.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.Network" Version="20.0.0-preview" />
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.Netcore.csproj]: undo this

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 2]
```diff
@@ -43,8 +43,8 @@
     <RunCodeAnalysis>false</RunCodeAnalysis>
   </PropertyGroup>
   <ItemGroup>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=19.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.4.0-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=20.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: undo this

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/NetworkProfileTests.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+using System;
+using Microsoft.Azure.ServiceManagemenet.Common.Models;
+using Microsoft.WindowsAzure.Commands.ScenarioTest;
+using Microsoft.WindowsAzure.Commands.Test.Utilities.Common;
+using Xunit;
+using Xunit.Abstractions;
+
+namespace Commands.Network.Test.ScenarioTests
+{
+    public class NetworkProfileTestsGenerated : RMTestBase
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/NetworkProfileTests.cs]: you need to checkin the test recordings for all these tests.

File: src/ResourceManager/Network/Commands.Network.Test/packages.config [hunk 4]
```diff
@@ -8,7 +8,8 @@
   <package id="Microsoft.Azure.Insights" version="0.13.1-preview" targetFramework="net45" />
   <package id="Microsoft.Azure.Management.Authorization" version="2.0.0" targetFramework="net452" />
   <package id="Microsoft.Azure.Management.Compute" version="21.0.0" targetFramework="net452" />
-  <package id="Microsoft.Azure.Management.Network" version="19.4.0-preview" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.ContainerInstance" version="1.1.0" targetFramework="net452" />
+  <package id="Microsoft.Azure.Management.Network" version="20.0.0-preview" targetFramework="net452" />
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network.Test/packages.config]: undo this change.. network library is at 19.4.0 for the latest SDK

File: src/ResourceManager/Network/Commands.Network/ChangeLog.md [hunk 5]
```diff
@@ -23,6 +23,11 @@
 
 ## Version 6.7.0
 * Updated cmdlet New-AzureRmVirtualNetworkGatewayConnection with support for switch ExpressRouteGatewayBypass
+* Added NetworkProfile functionality. New cmdlets added
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ChangeLog.md]: * Move this under `Current Version`
* There are more cmdlets added than listed here

File: src/ResourceManager/Network/Commands.Network/AzureRM.Network.psd1 [hunk 6]
```diff
@@ -338,7 +338,20 @@ CmdletsToExport = 'Add-AzureRmApplicationGatewayAuthenticationCertificate',
                'Get-AzureRmServiceEndpointPolicyDefinition', 
                'Set-AzureRmServiceEndpointPolicyDefinition', 
                'Add-AzureRmServiceEndpointPolicyDefinition', 
-               'Set-AzureRmServiceEndpointPolicy'
+               'Set-AzureRmServiceEndpointPolicy',
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/AzureRM.Network.psd1]: you ned to add the <Verb>-Az<Noun> version of these cmdlets to https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Network/Commands.Network/Az.Network.psd1

File: src/ResourceManager/Network/Commands.Network/Commands.Network.Netcore.csproj [hunk 7]
```diff
@@ -29,7 +29,7 @@
 
   <ItemGroup>
     <PackageReference Include="AutoMapper" Version="6.2.2" />
-    <PackageReference Include="Microsoft.Azure.Management.Network" Version="19.4.0-preview" />
+    <PackageReference Include="Microsoft.Azure.Management.Network" Version="20.0.0-preview" />
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.Netcore.csproj]: revert this

File: src/ResourceManager/Network/Commands.Network/Commands.Network.csproj [hunk 8]
```diff
@@ -58,8 +58,8 @@
       <HintPath>..\..\..\packages\Microsoft.Azure.KeyVault.Core.1.0.0\lib\net40\Microsoft.Azure.KeyVault.Core.dll</HintPath>
       <Private>True</Private>
     </Reference>
-    <Reference Include="Microsoft.Azure.Management.Network, Version=19.0.2.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
-      <HintPath>..\..\..\packages\Microsoft.Azure.Management.Network.19.4.0-preview\lib\net452\Microsoft.Azure.Management.Network.dll</HintPath>
+    <Reference Include="Microsoft.Azure.Management.Network, Version=20.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/Commands.Network.csproj]: revert this

**Comment by praries880** [comment]
@jcgknudson  
* Link the design review for the changes here
* Refer the design guidelines, pay attention to the [required param sets](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/azure-powershell-design-guidelines.md#required-parameter-sets) in it and add those wherever missing.
* Add test recording forthe tests

File: src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/NetworkProfileTests.cs (hunk 3 above)
Comment (jcgknudson) [reply src/ResourceManager/Network/Commands.Network.Test/ScenarioTests/NetworkProfileTests.cs]: yep, these will be in a subsequent iteration.

**Comment by praries880** [comment]
@jcgknudson  resolve the merge conflicts ASAP.

**Comment by praries880** [comment]
design review : https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/162

**Comment by jcgknudson** [comment]
> fix the build break in netcore : https://travis-ci.org/Azure/azure-powershell/builds/428280881#L1153
> 
> There might be double definition of a resource string.

Artifact of merge, my apologies. Fixing this now.

**Comment by praries880** [comment]
@jcgknudson  Most likely the [failures in the netcore build](https://travis-ci.org/Azure/azure-powershell/builds/428838901#L1979) are because of missing aliases use [this guide](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/create-alias-mappings.md) to create them.

File: src/ResourceManager/Network/Commands.Network/NetworkProfile/ContainerNetworkInterfaceConfiguration/NewAzureNetworkProfileContainerNetworkInterfaceConfigCommand.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "NetworkProfileContainerNicConfig", SupportsShouldProcess = true), OutputType(typeof(PSContainerNetworkInterfaceConfiguration))]
+    public partial class NewAzureNetworkProfileContainerNetworkInterfaceConfigCommand : AzureNetworkProfileContainerNetworkInterfaceConfigBase
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/NetworkProfile/ContainerNetworkInterfaceConfiguration/NewAzureNetworkProfileContainerNetworkInterfaceConfigCommand.cs]: where are Add-, Set-, REmove and Get- for this?

The [changes](https://github.com/Azure/azure-powershell/pull/7225/files#diff-22072398bfdba9fdc20e34ccc8051168R52) suggest that the following were added :
``` 
               'Get-AzureRmNetworkProfileContainerNicConfig', 
               'New-AzureRmNetworkProfileContainerNicConfig', 
               'Remove-AzureRmNetworkProfileContainerNicConfig', 
               'Add-AzureRmNetworkProfileContainerNicConfig', 
               'Set-AzureRmNetworkProfileContainerNicConfig', 
```

but I only see New

File: src/ResourceManager/Network/Commands.Network/NetworkProfile/IpConfigurationProfile/NewAzureNetworkProfileContainerNetworkInterfaceConfigIpConfigCommand.cs [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation;
+
+namespace Microsoft.Azure.Commands.Network
+{
+    [Cmdlet(VerbsCommon.New, ResourceManager.Common.AzureRMConstants.AzureRMPrefix + "NetworkProfileContainerNicConfigIpConfig", SupportsShouldProcess = true), OutputType(typeof(PSContainerNetworkInterfaceConfiguration))]
+    public partial class NewAzureNetworkProfileContainerNetworkInterfaceConfigIpConfigCommand : AzureNetworkProfileContainerNetworkInterfaceConfigIpConfigBase
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/NetworkProfile/IpConfigurationProfile/NewAzureNetworkProfileContainerNetworkInterfaceConfigIpConfigCommand.cs]: where are Add-, Set-, REmove and Get- for this?

The [changes](https://github.com/Azure/azure-powershell/pull/7225/files#diff-b60b2559dbc8dde377acf5fe189d4668R373) suggestthat the following were added :
``` 
               'Get-AzureRmNetworkProfileContainerNicConfigIpConfig', 
               'New-AzureRmNetworkProfileContainerNicConfigIpConfig', 
               'Add-AzureRmNetworkProfileContainerNicConfigIpConfig', 
               'Remove-AzureRmNetworkProfileContainerNicConfigIpConfig', 
               'Set-AzureRmNetworkProfileContainerNicConfigIpConfig', 
```

but I only see New

**Comment by praries880** [comment]
@jcgknudson 
You need to address the following :

AssemblyFileName | ClassName | Target | Severity | ProblemId | Description | Remediation
-- | -- | -- | -- | -- | -- | --
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Network\Microsoft.Azure.Commands.Network.dll | Microsoft.Azure.Commands.Network.RemoveAzureNetworkProfile | Remove-AzureRmNetworkProfile | 1 | 8510 | Cmdlet 'Remove-AzureRmNetworkProfile' has multiple parameter sets, but no   defined default parameter set. | Define a default parameter set in the cmdlet attribute.
c:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Network\Microsoft.Azure.Commands.Network.dll | Microsoft.Azure.Commands.Network.RemoveAzureNetworkProfile | Remove-AzureRmNetworkProfile | 1 | 8510 | Cmdlet 'Remove-AzureRmNetworkProfile' has multiple parameter sets, but no   defined default parameter set. | Define a default parameter set in the cmdlet attribute.

You will need to add a defauilt param set for any cmdlet that has more than one param set defined.

**Comment by praries880** [comment]
@jcgknudson  from the network test failures on desktop, it seems that the json recordings have still not been added to the desktop csproj, its causing the tests to fail : https://azuresdkci.westus2.cloudapp.azure.com/job/powershell/7877/artifact/src/Publish/TestResults/FailingTests/Microsoft.Azure.Commands.Network.Test.dll.html

File: src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs [hunk 11]
```diff
@@ -2253,7 +2253,19 @@ public class Mappings
     'New-AzNetworkInterfaceIpConfig': 'New-AzureRmNetworkInterfaceIpConfig',
     'New-AzExpressRouteCircuit': 'New-AzureRmExpressRouteCircuit',
     'Set-AzApplicationGatewayFrontendIPConfig': 'Set-AzureRmApplicationGatewayFrontendIPConfig',
-    'Get-AzInterfaceEndpoint': 'Set-AzureRmInterfaceEndpoint'
+    'Get-AzInterfaceEndpoint': 'Set-AzureRmInterfaceEndpoint',
```
Comment (praries880) [inline src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs]: you need to regenerate this file, as it still references the cmdlets that are not part of the release

**Comment by praries880** [comment]
@jcgknudson  you have a merge conflict

**Comment by jcgknudson** [comment]
> @jcgknudson you have a merge conflict

should be resolved

File: src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs (hunk 11 above)
Comment (jcgknudson) [reply src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs]: Not sure how this is done, then, I replaced the whole networking module just by running the CreateAliasMapping.ps1 script. Are there extra steps which are necessary?

File: src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs [hunk 12]
```diff
@@ -1928,336 +1928,368 @@ public class Mappings
...
-    'Add-AzRouteFilterRuleConfig': 'Add-AzureRmRouteFilterRuleConfig',
-    'Remove-AzLoadBalancerInboundNatPoolConfig': 'Remove-AzureRmLoadBalancerInboundNatPoolConfig',
-    'Get-AzVirtualNetworkGatewayConnectionVpnDeviceConfigScript': 'Get-AzureRmVirtualNetworkGatewayConnectionVpnDeviceConfigScript',
-    'Get-AzApplicationGatewayAvailableSslOptions': 'Get-AzureRmApplicationGatewayAvailableSslOptions',
-    'New-AzNetworkInterfaceIpConfig': 'New-AzureRmNetworkInterfaceIpConfig',
-    'New-AzExpressRouteCircuit': 'New-AzureRmExpressRouteCircuit',
-    'Set-AzApplicationGatewayFrontendIPConfig': 'Set-AzureRmApplicationGatewayFrontendIPConfig',
-    'Get-AzInterfaceEndpoint': 'Set-AzureRmInterfaceEndpoint'
-  },
+'Az.Network': {
```
[the remark this answers is not in the document]
Comment (jcgknudson) [reply src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs]: yes.. I may check out this change and add manually

[the remark this answers is not in the document]
Comment (maddieclayton) [reply src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs]: @jcgknudson That might be the easiest solution if you are having issues with the mapping script.  I am working on a change to make the manual generation unnecessary, but probably won't go in before your PR.

File: src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs [hunk 13]
```diff
@@ -556,9 +556,6 @@ public class Mappings
     'Remove-AzDataFactoryV2Pipeline': 'Remove-AzureRmDataFactoryV2Pipeline',
     'Remove-AzDataFactoryV2IntegrationRuntimeNode': 'Remove-AzureRmDataFactoryV2IntegrationRuntimeNode'
   },
-  'Az.ResourceGraph': {
-    'Search-AzGraph':  'Search-AzureRmGraph'
-  },
```
Comment (praries880) [inline src/ResourceManager/Profile/Commands.Profile/AzureRmAlias/Mappings.cs]: you need to add this back...

File: src/ResourceManager/Network/Commands.Network/ChangeLog.md [hunk 14]
```diff
@@ -44,6 +51,7 @@
   - Added New-AzureRmFirewallNatRule
   - Added New-AzureRmFirewallNetworkRuleCollection
   - Added New-AzureRmFirewallNetworkRule
+* Added service association link on Subnet Model
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network/ChangeLog.md]: move this line up under `Current Release` as well

**Comment by praries880** [comment]
on demand: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/1112/

PR Looks good to me, pending CI and on demand.

File: src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj [hunk 15]
```diff
@@ -352,6 +359,10 @@
     <None Include="SessionRecords\Commands.Network.Test.ScenarioTests.NetworkCloudExceptionTests\TestNotFound.json">
       <CopyToOutputDirectory>Always</CopyToOutputDirectory>
     </None>
+    <None Include="SessionRecords\Commands.Network.Test.ScenarioTests.NetworkProfileTests\TestContainerNetworkInterfaceConfigCRUD.json" />
```
Comment (praries880) [inline src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: @jdhuntington  the tests failed most likely because the entries are missing :
```
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory> 
    </None> 
```

the entry should be :
```
    <None Include="SessionRecords\Commands.Network.Test.ScenarioTests.NetworkProfileTests\TestContainerNetworkInterfaceConfigCRUD.json" > 
      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory> 
    </None> 
```
do the same for all the four json files

Comment (maddieclayton) [reply src/ResourceManager/Network/Commands.Network.Test/Commands.Network.Test.csproj]: fixed in github client - let's see if it passes now.

**Comment by praries880** [comment]
on demand passed, waiting for CI
